From 666b4e7005b1b538c0320fbf4117e4477a61f951 Mon Sep 17 00:00:00 2001
From: Julien Cristau <jcristau@debian.org>
Date: Mon, 12 Aug 2013 19:31:53 +0200
Subject: [PATCH 1/2] Link the driver against pixman

---
 configure.ac    |    1 +
 src/Makefile.am |    2 +-
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index 104113e..bd5ce9d 100644
--- a/configure.ac
+++ b/configure.ac
@@ -392,6 +392,7 @@ XORG_DRIVER_CHECK_EXT(DPMSExtension, xextproto)
 # Obtain compiler/linker options for the driver dependencies
 PKG_CHECK_MODULES(DRM, [libdrm >= 2.4.20]) # libdrm_intel is checked separately
 PKG_CHECK_MODULES(PCIACCESS, [pciaccess >= 0.10])
+PKG_CHECK_MODULES(PIXMAN, [pixman-1])
 
 LIBS=""
 AC_SEARCH_LIBS(clock_gettime, rt, [CLOCK_GETTIME_LIBS=$LIBS])
diff --git a/src/Makefile.am b/src/Makefile.am
index 16b6967..8810b74 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -37,7 +37,7 @@ AM_CFLAGS = \
 intel_drv_la_LTLIBRARIES = intel_drv.la
 intel_drv_la_LDFLAGS = -module -avoid-version
 intel_drv_ladir = @moduledir@/drivers
-intel_drv_la_LIBADD = legacy/liblegacy.la @PCIACCESS_LIBS@
+intel_drv_la_LIBADD = legacy/liblegacy.la @PCIACCESS_LIBS@ $(PIXMAN_LIBS)
 
 if SNA
 SUBDIRS += sna
-- 
1.7.10.4

