From cc9d7f6777703037d0d7eff346dd350f4d8b3f31 Mon Sep 17 00:00:00 2001
From: Chris Wilson <chris@chris-wilson.co.uk>
Date: Tue, 6 Jan 2015 13:29:49 +0000
Subject: sna: Reject CPU blits that require format conversions

Signed-off-by: Chris Wilson <chris@chris-wilson.co.uk>

diff --git a/src/sna/sna_render.c b/src/sna/sna_render.c
index 3fbb9ec..92c5afc 100644
--- a/src/sna/sna_render.c
+++ b/src/sna/sna_render.c
@@ -2336,6 +2336,9 @@ memcpy_copy_boxes(struct sna *sna, uint8_t op,
 	if (op != GXcopy)
 		return false;
 
+	if (src_draw->depth != dst_draw->depth)
+		return false;
+
 	clipped = (n > 1 ||
 		   box->x1 + dx > 0 ||
 		   box->y1 + dy > 0 ||
-- 
cgit v0.10.2

